### PR TITLE
cmd/k8s-operator: fix base truncating for extra long Service names

### PR DESCRIPTION
**Pull request by irbekrm**
#### Context

StatefulSet names for ingress/egress proxies [are calculated using Kubernetes name generator and the parent resource name as a base](https://github.com/tailscale/tailscale/blob/v1.56.1/cmd/k8s-operator/sts.go#L215-L218).
We currently try pre-calculate how long the generated name will be and cut the base it is to be longer than 52 chars, see https://github.com/tailscale/tailscale/pull/10343.

#### Problem

The name generator also cuts the base, but has [a higher max cap](https://github.com/kubernetes/kubernetes/blob/v1.28.4/staging/src/k8s.io/apiserver/pkg/storage/names/generate.go#L46). This commit fixes a bug where, if we get a shortened base back from _the generator_, we cut off too little as the base that we have cut will be passed into the generator again, which will then itself cut less because the base is shorter- so we end up with a too long name again.

#### Fix

This PR fixes it by cutting recursively to ensure that, if the cut base produces a longer name again after being put through the name  generator, we cut again till we ensure that the generated name will definitely be no longer than 52 chars.

#### Notes

The longest `Service` name supported before this fix is 54 chars. I have tested that with the contents of this PR vs latest release and verified that both produced the same base (so that the change in this PR does not calculate different names for existing StatefulSets)

Updates tailscale/tailscale#10807

### CODE REVIEW

File: cmd/k8s-operator/sts.go [hunk 1]
```diff
@@ -214,16 +214,28 @@ const maxStatefulSetNameLength = 63 - 10 - 1
...
-		base = base[:len(base)-excess-1] // take extra char off to make space for hyphen
-		base = base + "-"                // re-instate hyphen
+		base = base[:len(base)-1-excess] // cut off the excess chars
+	}
+	if !strings.HasSuffix(base, "-") { // dash may have been cut by the generator
+		base = base + "-"
+	}
+	generatedName = generator.GenerateName(base)
+	if excess := len(generatedName) - maxStatefulSetNameLength; excess > 0 {
+		return redactBase(base)
```
[the remark this answers is not in the document]
Comment (irbekrm) [reply cmd/k8s-operator/sts.go]: Thanks @maisem I've fixed that.

File: cmd/k8s-operator/sts.go [hunk 2]
```diff
@@ -214,16 +214,21 @@ const maxStatefulSetNameLength = 63 - 10 - 1
...
+		}
+		if !strings.HasSuffix(base, "-") { // dash may have been cut by the generator
+			base = base + "-"
+		}
+		generatedName = generator.GenerateName(base)
+		if excess := len(generatedName) - maxStatefulSetNameLength; excess <= 0 {
+			tooLong = false
+		}
 	}
 	return base
```
Comment (maisem) [inline cmd/k8s-operator/sts.go:233]:
```suggestion
	for {
		generatedName := generator.GenerateName(base)
		excess := len(generatedName) - maxStatefulSetNameLength
		if excess <= 0 {
			return base
		}
		base = base[:len(base)-1-excess] // cut off the excess chars
		if !strings.HasSuffix(base, "-") { // dash may have been cut by the generator
			base = base + "-"
		}
	}
```
?

File: cmd/k8s-operator/sts_test.go [hunk 3]
```diff
@@ -19,32 +22,21 @@ import (
...
+	for len(b.String()) < 63 {
+		baseLength := len(b.String())
+		if baseLength > 43 {
+			baseLength = 43 // currently 43 is the max base length
+		}
+		wantsNameR := regexp.MustCompile(`^ts-a{` + fmt.Sprint(baseLength) + `}-$`) // to match a string like ts-aaaa-
+		gotName := statefulSetNameBase(b.String())
+		if _, err := b.WriteString("a"); err != nil {
+			t.Fatalf("error writing to string builder: %v", err)
+		}
```
Comment (maisem) [inline cmd/k8s-operator/sts_test.go:37]: move this to the top of the loop and remove line 27. that way its clear that you are starting with `a` and adding `a` every loop?

File: cmd/k8s-operator/sts_test.go [hunk 4]
```diff
@@ -19,32 +22,21 @@ import (
...
-		t.Run(tt.name, func(t *testing.T) {
-			if got := statefulSetNameBase(tt.in); got != tt.out {
-				t.Errorf("stsNamePrefix(%s) = %q, want %s", tt.in, got, tt.out)
-			}
-		})
+	// Service name lengths can be 1 - 63 chars, be paranoid and test them all.
+	var b strings.Builder
+	b.WriteString("a")
+	for len(b.String()) < 63 {
+		baseLength := len(b.String())
```
Comment (maisem) [inline cmd/k8s-operator/sts_test.go:29]: b.Len()

**Review by maisem — APPROVED** [review]

File: cmd/k8s-operator/sts.go (hunk 2 above)
Comment (irbekrm) [reply cmd/k8s-operator/sts.go:233]: ooh a lot nicer

**Pull request merged**